src/utils.jl

```diff
@@ -84,16 +85,27 @@ function build_layer_from_vec!(
   return sizecuArray
 end
 
+function build_layer_from_vec!(
+  cuArray::Array{T, N} where {N},
+  v::Vector{T},
+  index::Int,
+) where {T <: Number}
+  sizecuArray = reduce(*, size(cuArray))
+  copyto!(cuArray, v[(index + 1):(index + sizecuArray)])
+  return sizecuArray
+end
+
 """
     build_nested_array_from_vec(chain_ANN :: C, v :: Vector{T}) where {C <: Chain, T <: Number}
-    build_nested_array_from_vec(model :: KnetNLPModel{T, S, C}, v :: Vector{T}) where {T, S, C}
+    build_nested_array_from_vec(model :: KnetNLPModel{T, S, C, V}, v :: Vector{T}) where {T, S, C, V}
     build_nested_array_from_vec(nested_array :: Vector{CuArray{T, N, CUDA.Mem.DeviceBuffer} where N}, v :: Vector{T}) where {T <: Number}
+    build_nested_array_from_vec(nested_array :: Vector{Array{T,N} where N}, v :: Vector{T}) where {T <: Number}
```

Comment:
Could this be `AbstractVector{A} where A <: AbstractArray{T,N}`?

Reply:
Same for other methods.

Reply:
It does not work.
I implemented it, in the application it is said te signatures don't match.
```julia
KnetNLPModels tests: Error During Test at C:\Users\raynaudp\Desktop\these\code\KnetNLPModels.jl\test\runtests.jl:6
  Got exception outside of a @test
  MethodError: no method matching build_nested_array_from_vec!(::Vector{Array{Float32}}, ::Vector{Float32})
  Closest candidates are:
    build_nested_array_from_vec!(::KnetNLPModel{T, S, C, V}, ::Vector{T}) where {T, S, C, V} at C:\Users\raynaudp\Desktop\these\code\KnetNLPModels.jl\src\utils.jl:150
    build_nested_array_from_vec!(::AbstractVector{A}, ::Vector{T}) where {T<:Number, N, A<:AbstractArray{T, N}} at C:\Users\raynaudp\Desktop\these\code\KnetNLPModels.jl\src\utils.jl:153
```

Reply:
The approach `AbstractVector{A} where A <: AbstractArray{T,N}` doesn't work.
It does not recongnize the `Vector{Array{T,N} where N}` or the `Vector{CuArray{T, N, CUDA.Mem.DeviceBuffer} where N}`.
```julia
KnetNLPModels tests: Error During Test at C:\Users\raynaudp\Desktop\these\code\KnetNLPModels.jl\test\runtests.jl:6
  Got exception outside of a @test
  MethodError: no method matching build_nested_array_from_vec!(::Vector{Array{Float32}}, ::Vector{Float32})
  Closest candidates are:
    build_nested_array_from_vec!(::KnetNLPModel{T, S, C, V}, ::Vector{T}) where {T, S, C, V} at C:\Users\raynaudp\Desktop\these\code\KnetNLPModels.jl\src\utils.jl:150
    build_nested_array_from_vec!(::AbstractVector{A}, ::Vector{T}) where {T<:Number, N, A<:AbstractArray{T, N}} at C:\Users\raynaudp\Desktop\these\code\KnetNLPModels.jl\src\utils.jl:153
```
I implemented it as well as `(nested_array::B) where {T, N, B<:AbstractVector{A}, A <:AbstractArray{T,N}`with the same result.

Reply:
This works:
```julia
julia> function bla(x::AbstractVector{<: AbstractArray})
           println("in bla!")
       end

julia> a = [rand(2), rand(2,8)];

julia> bla(a)
in bla!

julia> function bla(x::AbstractVector{<: AbstractArray{T}}) where T<:Number
           println("in bla for numbers!")
       end

julia> bla(a)
in bla for numbers!

julia> b = [["1", "2"], [:x, π]];

julia> bla(b)
in bla!
```
Is your situation different?

Reply:
I tried your code, with a `map` inside to close the gap with the methods in KnetNLPModels.jl
```julia
julia> function bla(x::V) where {N, T<:Number,  M<:AbstractArray{T, N}, V<:AbstractVector{M}}
           map(i -> println("the "* string(i) *"-th component is of size" * string(size(x[i]))), 1:length(x))
       end
julia> a = [rand(2), rand(2,8), rand(4,5,6)]
julia> bla(a)
the 1-th component is of size(2,)
the 2-th component is of size(2, 8)
the 3-th component is of size(4, 5, 6)
```
It is working fine.
I tried to implement 
```julia
function build_nested_array_from_vec(
  nested_array::AbstractVector{<:AbstractArray{T,N}},
  v::Vector{T},
) where {T <: Number, N}
  vec_cuarray = map(i -> similar(nested_array[i]), 1:length(nested_array))
  build_nested_array_from_vec!(vec_cuarray, v)
  return vec_cuarray
end
```
But i ended up with an error.

I move the `N` as:
```julia
function build_nested_array_from_vec(
  nested_array::AbstractVector{<:AbstractArray{T,N} where {N}},
  v::Vector{T},
) where {T <: Number}
  vec_cuarray = map(i -> similar(nested_array[i]), 1:length(nested_array))
  build_nested_array_from_vec!(vec_cuarray, v)
  return vec_cuarray
end
```
an it worked.

Reply:
I'm recoding the functions involved by this change.

Reply:
Well, I did not know that it was legal to put a `where N` at that place! 👍